robot/impl/resource_manager.go

```diff
@@ -537,115 +539,150 @@ func (manager *resourceManager) completeConfig(
+
+					// this is done in config validation but partial start rules require us to check again
+					if _, err := conf.Validate("", resName.API.Type.Name); err != nil {
+						gNode.LogAndSetLastError(
+							fmt.Errorf("resource config validation error: %w", err),
+							"resource", conf.ResourceName(),
+							"model", conf.Model)
+						return
+					}
+					if manager.moduleManager.Provides(conf) {
+						if _, err := manager.moduleManager.ValidateConfig(ctxWithTimeout, conf); err != nil {
+							gNode.LogAndSetLastError(
+								fmt.Errorf("modular resource config validation error: %w", err),
+								"resource", conf.ResourceName(),
+								"model", conf.Model)
+							return
+						}
+					}
 
-			switch {
-			case resName.API.IsComponent(), resName.API.IsService():
-				newRes, newlyBuilt, err := manager.processResource(ctxWithTimeout, conf, gNode, lr)
-				if newlyBuilt || err != nil {
-					if err := manager.markChildrenForUpdate(resName); err != nil {
-						manager.logger.CErrorw(ctx,
-							"failed to mark children of resource for update",
-							"resource", resName,
-							"reason", err)
+					switch {
+					case resName.API.IsComponent(), resName.API.IsService():
+
+						newRes, newlyBuilt, err := manager.processResource(ctxWithTimeout, conf, gNode, lr)
+						if newlyBuilt || err != nil {
+							if err := manager.markChildrenForUpdate(resName); err != nil {
+								manager.logger.CErrorw(ctx,
+									"failed to mark children of resource for update",
+									"resource", resName,
+									"reason", err)
+							}
+						}
+
+						if err != nil {
+							gNode.LogAndSetLastError(
+								fmt.Errorf("resource build error: %w", err),
+								"resource", conf.ResourceName(),
+								"model", conf.Model)
+							return
+						}
+
+						// if the ctxWithTimeout fails with DeadlineExceeded, then that means that
+						// resource generation is running async, and we don't currently have good
+						// validation around how this might affect the resource graph. So, we avoid
+						// updating the graph to be safe.
+						if errors.Is(ctxWithTimeout.Err(), context.DeadlineExceeded) {
+							manager.logger.CErrorw(
+								ctx, "error building resource", "resource", conf.ResourceName(), "model", conf.Model, "error", ctxWithTimeout.Err())
+						} else {
+							gNode.SwapResource(newRes, conf.Model)
+						}
+
+					default:
+						err := errors.New("config is not for a component or service")
+						gNode.LogAndSetLastError(err, "resource", resName)
 					}
-				}
+				})
 
-				if err != nil {
-					gNode.LogAndSetLastError(
-						fmt.Errorf("resource build error: %w", err),
-						"resource", conf.ResourceName(),
-						"model", conf.Model)
-					return
+				select {
+				case <-resChan:
+				case <-ctxWithTimeout.Done():
+					if errors.Is(ctxWithTimeout.Err(), context.DeadlineExceeded) {
+						lr.logger.CWarn(ctx, rutils.NewBuildTimeoutError(resName.String()))
+					}
+				case <-ctx.Done():
```

Comment:
Dumb question: I know you didn't write this but:

`ctxWithTimeout` is forked from `ctx`. Do we actually need to listen to both channels here or is `ctxWithTimeout` sufficient?

And if `ctxWithTimeout` does "speak for" `ctx`, I suppose we have a bug today where `cancel(ctx)` may or may not result in an error being returned in this select? Based on which channel the select statement observes first?